File: hw4/python/task2.py
import os
import sys
import time
import copy
from itertools import combinations

from pyspark import SparkContext

class Node:
    def __init__(self, name):
        self.name = name
        self.value = 1
        self.parents = set()
        self.children = set()

def get_communities(nodes, edges):
    visited_nodes = set()
    communities = list()
    for n in nodes:
        if len(visited_nodes) == len(nodes):
            break
        elif n in visited_nodes:
            continue
        else:
            community = set()
            candidate_nodes = {n}
            while len(candidate_nodes) > 0:
                for nn in copy.deepcopy(candidate_nodes):
                    for e in copy.deepcopy(edges):
                        if nn in e:
                            if e[0] not in visited_nodes:
                                candidate_nodes.add(e[0])
                                visited_nodes.add(e[0])
                                community.add(e[0])
                            if e[1] not in visited_nodes:
                                candidate_nodes.add(e[1])
                                visited_nodes.add(e[1])
                                community.add(e[1])
                            edges.remove(e)
                    candidate_nodes.remove(nn)
            if len(community) == 0: communities.append({n})
            else: communities.append(community)
    return communities


def calculate_graph_betweenness(communities, edges):
    edge_betweenness = dict()
    for c in communities:
        edges_dict = dict()
        c_edges = set()
        for edge in copy.deepcopy(edges):
            if edge[0] in c and edge[1] in c:
                c_edges.add(edge)
                edges.remove(edge)
        for e in c_edges:
            edges_dict.setdefault(e[0], []).append(e[1])
            edges_dict.setdefault(e[1], []).append(e[0])

        for node in c:
            betweenness = get_tree_betweenness(node, copy.deepcopy(c_edges), copy.deepcopy(edges_dict))
            for item in betweenness:
                edge_betweenness[item[0]] = edge_betweenness.setdefault(item[0], 0) + item[1]
    return list(map(lambda x: (x[0], x[1] / 2), edge_betweenness.items()))

def get_tree_betweenness(root, edges, edges_dict):
    root_node = Node(root)
    level_dict = dict()
    level = 0
    level_dict[level] = {root_node}
    node_dict = {root_node.name: root_node}
    while len(level_dict[level]) > 0:
        level += 1
        level_dict[level] = set()
        for node in level_dict[level - 1]:
            children = set([(node_dict[x] if x in node_dict.keys() else Node(x)) for x in edges_dict[node.name]])
            for c in children:
                node_dict[c.name] = c
                edges_dict[node.name].remove(c.name)
                edges_dict[c.name].remove(node.name)
                if c in level_dict[level - 1]:
                    continue
                node.children.add(c)
                c.parents.add(node)
                level_dict[level].add(c)

    edges_betweenness = tree_betweenness(edges, level_dict)
    return list(edges_betweenness.items())

def tree_betweenness(edges, level_dict):
    nodes_betweeness = dict()
    edges_betweenness = dict(map(lambda x:(x, 0), edges))
    levels = sorted(level_dict.items(), key=lambda x:x[0], reverse=True)
    for level in levels:
        for node in level[1]:
            if len(node.children) == 0:
                nodes_betweeness[node.name] = 1
            else:
                children = node.children
                children_links = [tuple(sorted([node.name, child.name])) for child in children]
                nodes_betweeness[node.name] = 1 + sum([edges_betweenness[link] for link in children_links])
            if len(node.parents) != 0:
                if len(list(node.parents)[0].parents) > 0:
                    weights_dict = dict([(n, len(n.parents)) for n in node.parents])
                else:
                    weights_dict = dict([(n, 1) for n in node.parents])
                score = 1 / sum(weights_dict.values()) * nodes_betweeness[node.name]
                for p in node.parents:
                    link = tuple(sorted([node.name, p.name]))
                    edges_betweenness[link] += score * weights_dict[p]
    return edges_betweenness

def generate_tree(root_node, edges_dict):
    level_dict = dict()
    level = 0
    level_dict[level] = {root_node}
    node_dict = {root_node.name: root_node}
    while len(level_dict[level]) > 0:
        level += 1
        level_dict[level] = set()
        for node in level_dict[level-1]:
            children = set([(node_dict[x] if x in node_dict.keys() else Node(x)) for x in edges_dict[node.name]])
            for c in children:
                node_dict[c.name] = c
                edges_dict[node.name].remove(c.name)
                edges_dict[c.name].remove(node.name)
                if c in level_dict[level-1]:
                    continue
                node.children.add(c)
                c.parents.add(node)
                level_dict[level].add(c)
    return level_dict


def calculate_modularity(commuities, m, nodes_degree, adjacent_dict):
    modularity = 0
    for c in commuities:
        tmp_mod = 0

        for pair in combinations(c, 2):
            pair = tuple(sorted(pair))
            aij = adjacent_dict[pair]
            tmp_mod += aij - nodes_degree[pair[0]] * nodes_degree[pair[1]] / 2 / m
        modularity += tmp_mod
    return modularity / 2 / m

if __name__ == '__main__':
    start_time = time.time()
    os.environ['JAVA_HOME'] = "/home/buyi/app/jdk"
    os.environ['PYSPARK_PYTHON'] = '/usr/bin/python3'

    filter_threshold, input_file_path, betweenness_output_file_path, community_output_file_path = '7', '/home/buyi/data/hw4/ub_sample_data.csv', 'task2_1_ans', 'task2_2_ans'
    # filter_threshold, input_file_path, betweenness_output_file_path, community_output_file_path = sys.argv[1:]
    sc = SparkContext("local[3]", "task2")

    rdd = sc.textFile(input_file_path)
    header = rdd.first()
    rdd1 = rdd.filter(lambda x:x != header).map(lambda x:x.split(','))
    user_index = rdd1.map(lambda x:x[0]).distinct().zipWithIndex().collectAsMap()
    index_user = dict([(x[1], x[0]) for x in user_index.items()])
    business_index = rdd1.map(lambda x:x[1]).distinct().zipWithIndex().collectAsMap()
    rdd_user_profile = rdd1.map(lambda x:(user_index[x[0]],business_index[x[1]])).groupByKey().mapValues(lambda x:list(x))
    user_profile_dict = rdd_user_profile.collectAsMap()
    unique_user = user_index.values()
    pairs = combinations(unique_user, 2)
    nodes, edges = set(), set()
    for p in pairs:
        if len(set(user_profile_dict[p[0]]).intersection(user_profile_dict[p[1]])) >= int(filter_threshold):
            nodes.add(p[0])
            nodes.add(p[1])
            edges.add(tuple(sorted(p)))
    # [{38, 2, 213, 6},{584, 1060}, ...]
    communities = get_communities(nodes, copy.deepcopy(edges))

    betweenness_edges = calculate_graph_betweenness(communities, copy.deepcopy(edges))
    sorted_betweenness_edges = sorted(betweenness_edges, key=lambda x:x[1], reverse=True)

    with open(betweenness_output_file_path, 'w') as f:
        betweenness_edges = list(map(lambda x:(tuple(sorted([index_user[x[0][0]], index_user[x[0][1]]])), x[1]), betweenness_edges))
        for pair_betweenness in sorted(sorted(betweenness_edges, key=lambda x:x[0]), key=lambda x:x[1], reverse=True):
            f.write("('{}', '{}'), {}\n".format(pair_betweenness[0][0], pair_betweenness[0][1], pair_betweenness[1]))

    nodes_degree = dict()
    for e in edges:
        nodes_degree[e[0]] = nodes_degree.setdefault(e[0], 0) + 1
        nodes_degree[e[1]] = nodes_degree.setdefault(e[1], 0) + 1
    adjacent_dict = dict()
    for pair in combinations(nodes_degree.keys(), 2):
        pair = tuple(sorted(pair))
        adjacent_dict[pair] = 1 if pair in edges else 0
    modularity, max_modularity = -1, -1
    m = len(sorted_betweenness_edges)
    pre_communities = copy.deepcopy(communities)
    while modularity >= max_modularity:
        pre_communities = copy.deepcopy(communities)
        max_modularity = modularity
        highest_betweenness = sorted_betweenness_edges[0][1]
        for e in copy.deepcopy(sorted_betweenness_edges):
            if e[1] == highest_betweenness:
                sorted_betweenness_edges.remove(e)
        edges = list(map(lambda x: x[0], sorted_betweenness_edges))
        communities = get_communities(nodes, copy.deepcopy(edges))

        modularity = calculate_modularity(communities, m, nodes_degree, adjacent_dict)
        betweenness_edges = calculate_graph_betweenness(communities, edges)
        sorted_betweenness_edges = sorted(betweenness_edges, key=lambda x:x[1], reverse=True)
    sorted_communities = sorted(sorted(list(map(lambda x:sorted([index_user[i] for i in x]), pre_communities))), key=lambda x:len(x))
    # print(sorted_communities)
    with open(community_output_file_path, 'w') as f:
        for c in sorted_communities:
            f.write(','.join(["'"+x+"'" for x in c])+'\n')
    print("time: {}".format(time.time() - start_time))

File: hw4/scala/task2.scala
import org.apache.spark.{SparkConf, SparkContext}

import java.io.{File, PrintWriter}
import java.util.Date
import scala.collection.mutable
import scala.util.control.Breaks._

class Node(na: Int){
  var name: Int = na
  var value: Double = 1.0
  var parents: Set[Node] = Set()
  var children: Set[Node] = Set()
}

object task2 {

  def get_communities(nodes: Set[Int], edges: Set[(Int, Int)]): Seq[Set[Int]] = {
    var visited_nodes:Set[Int] = Set()
    var communities:List[Set[Int]] = List()
    var tmp_edges = edges
    breakable {
      for (n <- nodes) {
        if (visited_nodes.size == nodes.size) break
        else if (!visited_nodes.contains(n)) {
          var community:Set[Int] = Set()
          var candidate_nodes:Set[Int] = Set(n)
          while (candidate_nodes.nonEmpty) {
            val tmp_candidate = candidate_nodes
            for (nn <- tmp_candidate) {
              for (e <- tmp_edges) {
                if (List(e._1, e._2).contains(nn)){
                  for (p <- List(e._1, e._2)){
                    if (!visited_nodes.contains(p)){
                      candidate_nodes += p
                      visited_nodes += p
                      community += p
                    }
                  }
                  tmp_edges -= e
                }
              }
              candidate_nodes -= nn
            }
          }
          if (community.isEmpty) communities = communities :+ Set(n)
          else communities = communities :+ community
        }
      }
    }
    communities
  }

  def tree_betweenness(c_edges:Set[(Int, Int)], level_dict: Map[Int, mutable.Set[Node]]): Map[(Int, Int), Double] = {
    var nodes_betweenness:Map[Int, Double] = Map()
    var edges_betweenness:Map[(Int, Int), Double] = Map()
    c_edges.foreach(item=>edges_betweenness += (item -> 0.0))
//    edges.foreach(item=>edges_betweenness += (item -> 0.0))
    val levels = level_dict.keys.toList.sorted.reverse
    for (level <- levels) {
      for (node <- level_dict.getOrElse(level, mutable.Set())){
        val children = node.children
        if (children.isEmpty) {nodes_betweenness += (node.name -> 1)}
        else {
          val children_links = children.map(child=>(node.name, child.name)).map(item=>List(item._1, item._2).sorted)
            .map(item=>(item.head, item.apply(1))).toList
//          println(children_links.head, children_links.apply(1))
//          val chil = children_links.map(link=> edges_betweenness.get(link).head)
//          val tmp_s = 1+ children_links.map(link=> edges_betweenness.get(link).head).sum
          nodes_betweenness += (node.name -> (1+ children_links.map(link=> edges_betweenness.get(link).head).sum))
//          println("kkkkkkk", nodes_betweenness.getOrElse(node.name, 0))
        }

        if (node.parents.nonEmpty) {
          var weights_dict:Map[Node, Int] = Map()
          if (node.parents.head.parents.nonEmpty) node.parents.foreach(n=>{
            weights_dict += (n -> n.parents.size)
          }) else node.parents.foreach(n=>weights_dict += (n -> 1))
          val score:Double = (1.0 / weights_dict.values.sum) * nodes_betweenness.getOrElse(node.name, 0.0)
//          println(score, weights_dict.values.sum, nodes_betweenness.getOrElse(node.name, 0.0))
//          println(1.0 / weights_dict.values.sum, nodes_betweenness.getOrElse(node.name, 0.0), score)
          for (p <- node.parents) {
            val link = List(node.name, p.name).sorted
//            if(link.head == 2 && link.apply(1) == 5) println(nodes_betweenness)
            edges_betweenness += ((link.head, link.apply(1)) -> score * weights_dict.getOrElse(p, 0))
          }
        }
      }
    }
//    println(edges_betweenness)
    edges_betweenness
  }

  def get_tree_betweenness(node: Int, c_edges:Set[(Int, Int)], edges_dict: Map[Int, mutable.Set[Int]]): Map[(Int, Int), Double] = {
    var root_node = new Node(node)
    var level = 0
    var level_dict:Map[Int, mutable.Set[Node]] = Map(level -> mutable.Set(root_node))

    var node_dict = Map(root_node.name->root_node)
    var tmp_edges_dict: Map[Int, mutable.Set[Int]] = Map()
//    edges_dict.foreach(item=>tmp_edges_dict += (item._1 -> item._2:_*))
    for (k <- edges_dict.keys) {tmp_edges_dict += (k -> mutable.Set(edges_dict.getOrElse(k, mutable.Set()).toSeq:_*))}
    while (level_dict.getOrElse(level, mutable.Set()).nonEmpty) {
      level += 1
      level_dict += (level -> mutable.Set())
      for (node <- level_dict.getOrElse(level-1, mutable.Set())) {
        var children:Set[Node] = Set()
        tmp_edges_dict.getOrElse(node.name, mutable.Set()).foreach(x=>{
          if (!node_dict.contains(x)) node_dict += (x -> new Node(x))
          children += node_dict.get(x).head
        })
//        println(level, children.map(_.name))
        for (c <- children) {
          breakable {
            node_dict += (c.name -> c)
            tmp_edges_dict += (node.name -> (tmp_edges_dict.getOrElse(node.name, mutable.Set()) -= c.name))
            tmp_edges_dict += (c.name -> (tmp_edges_dict.getOrElse(c.name, mutable.Set()) -= node.name))
            if (level_dict.getOrElse(level-1, mutable.Set()).contains(c)) break
            node.children += c
            c.parents += node
            level_dict += (level -> (level_dict.getOrElse(level, mutable.Set()) += c))
          }
        }
      }
    }
//    for (item <- level_dict){
//      println(item._1, item._2.map(item=>(item.name,item.parents.map(_.name))))
//    }
//    println(c_edges)
    tree_betweenness(c_edges, level_dict)
  }

  def calculate_graph_betweenness(communities: Seq[Set[Int]], edges: Set[(Int, Int)]): Map[(Int, Int), Double] = {
    var edge_betweenness: Map[(Int, Int), Double] = Map()
    for (c <- communities) {
      var edges_dict:Map[Int, mutable.Set[Int]]= Map()
      var c_edges:Set[(Int, Int)] = Set()
      var tmp_edges = edges
      for (edge <- tmp_edges){
        if (c.contains(edge._1) || c.contains(edge._2)){
          c_edges += edge
          tmp_edges -= edge
        }
      }
      for (e<- c_edges) {
        edges_dict += (e._1 -> (edges_dict.getOrElse(e._1, mutable.Set()) + e._2))
        edges_dict += (e._2 -> (edges_dict.getOrElse(e._2, mutable.Set()) + e._1))
      }
      for (node <- c) {
        val betweenness: Map[(Int, Int), Double] = get_tree_betweenness(node, c_edges, edges_dict)

        for (item <- betweenness) {
          edge_betweenness += (item._1 -> (edge_betweenness.getOrElse(item._1, 0.0)+ item._2) )
        }
      }

    }
    edge_betweenness.map(item=>(item._1, item._2 / 2))
  }

  def calculate_modularity(communities: Seq[Set[Int]], m: Int, nodes_degree: Map[Int, Int], adjacent_dict: Map[(Int, Int), Int]): Double = {
    var modularity = 0.0
    for (c <- communities) {
      var tmp_m = 0.0
      for (pair <- c.toList.combinations(2)){
        var p = (pair.min, pair.max)
        val aij = adjacent_dict.getOrElse(p, 0)
        tmp_m += aij - nodes_degree.getOrElse(p._1, 0) * nodes_degree.getOrElse(p._2,0) / 2.0 / m
      }
      modularity += tmp_m
    }
    modularity / 2.0 / m
  }

  def main(args: Array[String]): Unit = {
    val start_time = new Date().getTime
    val conf = new SparkConf().setMaster("local[3]").setAppName("task1")
    val sc = new SparkContext(conf)
    sc.setLogLevel("WARN")

    val filter_threshold = args(0) //"7"
    val input_file_path = args(1) //"E:\\Data\\inf553\\hw4\\ub_sample_data.csv"
    val betweenness_output_file_path = args(2) //"task2_1_scala"
    val community_output_file_path = args(3) //"task2_2_scala"

    val rdd = sc.textFile(input_file_path)
    val header = rdd.first()
    val rdd1 = rdd.filter(item=>item!=header).map(_.split(",")).map(item=>Tuple2(item.head, item.apply(1)))
    val user_index = rdd1.map(_._1).distinct().zipWithIndex().collectAsMap()
    val index_user = user_index.map(item=>item._2->item._1)
    val business_index = rdd1.map(_._2).distinct().zipWithIndex().collectAsMap()
    val user_profile_dict = rdd1.map(item=>(user_index.get(item._1).head, business_index.get(item._2).head))
      .groupByKey().mapValues(_.toSeq).collectAsMap()
    val unique_user_index = sc.parallelize(index_user.keys.toSeq)
    var nodes:Set[Int] = Set()
    var edges:Set[(Int, Int)] = Set()
    unique_user_index.cartesian(unique_user_index).filter(item=>item._1 < item._2)
      .filter(item=>(user_profile_dict.get(item._1).head.toSet.intersect(user_profile_dict.get(item._2).head.toSet)).size >= filter_threshold.toInt)
      .collect().foreach(item=>{
      nodes += item._1.toInt
      nodes += item._2.toInt
      edges += ((item._1.toInt, item._2.toInt))
    })
    var communities = get_communities(nodes, edges)
    var betweenness_edges = calculate_graph_betweenness(communities, edges) //val betweenness_edges =
//    println(betweenness_edges.toList.sortBy(_._2).reverse)
    val pw = new PrintWriter(new File(betweenness_output_file_path))
    var sorted_betweenness_edges = betweenness_edges.map(item=>(List(index_user.getOrElse(item._1._1, ""), index_user.getOrElse(item._1._2, "")), item._2))
      .map(item=>((item._1.min, item._1.max), item._2)).toList.sortBy(_._2).reverse
    sorted_betweenness_edges.foreach(item=>pw.write("('"+item._1._1+"', '"+item._1._1+"'), "+item._2.toString+"\n"))
    pw.close()

    var nodes_degree:Map[Int, Int] = Map()
    for (e<-betweenness_edges.keys) {
      nodes_degree += (e._1 -> (nodes_degree.getOrElse(e._1, 0) + 1))
      nodes_degree += (e._2 -> (nodes_degree.getOrElse(e._2, 0) + 1))
    }
//    println(nodes_degree)
    var adjacent_dict:Map[(Int,Int),Int] = Map()
    for (pair <- nodes_degree.keys.toList.combinations(2)){
      val tmp = (pair.min, pair.max)
      var value = 0
      if (edges.contains(tmp)) adjacent_dict += (tmp -> 1)
    }
    var modularity = -1.0
    var max_modularity = -1.0
    val m = edges.size
//    println(m)
    var pre_communities = communities
    var sorted_dict = betweenness_edges.toList.sortBy(_._2).reverse
    while (modularity >= max_modularity) {
      pre_communities = communities
      max_modularity = modularity
//      println(communities.size, modularity, max_modularity)
      val highest_betweenness = sorted_dict.head._2
//      var copy_sorted = sorted_betweenness_edges
      sorted_dict = sorted_dict.filter(item=>item._2!=highest_betweenness)
      edges = sorted_dict.map(_._1).toSet
      communities = get_communities(nodes, edges)
      modularity =  calculate_modularity(communities, m, nodes_degree, adjacent_dict)
      betweenness_edges = calculate_graph_betweenness(communities, edges)
      sorted_dict = betweenness_edges.toList.sortBy(_._2).reverse
    }
    val pw1 = new PrintWriter(new File(community_output_file_path))
    val sorted_communities = pre_communities.map(_.map(item=>index_user.getOrElse(item, "")).toList.sorted).sortBy(_.head).sortBy(_.size)
    sorted_communities.map(item=>item.map(i=>"'"+i+"'")).foreach(item=>pw1.write(item.mkString(",")+"\n"))
    pw1.close()
    println("time: "+(new Date().getTime-start_time)/1000.0+"s")
  }
}
